### PR TITLE
refactor: switch to new gateway-api interfaces

### PR DESCRIPTION
**Pull request by jhaeyaert**
## Issue

https://gravitee.atlassian.net/browse/APIM-7109

## Description

This PR prepares for native Kafka API and switches to the new gateway-api interfaces internally but keeps backward compatibility.

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-dviqwdjnbl.chromatic.com)

### CODE REVIEW

File: gravitee-apim-gateway/gravitee-apim-gateway-flow/src/main/java/io/gravitee/gateway/reactive/flow/AbstractFlowResolver.java [hunk 1]
```diff
@@ -17,22 +17,24 @@
...
+import io.gravitee.gateway.reactive.api.context.http.HttpBaseExecutionContext;
 import io.gravitee.gateway.reactive.core.condition.ConditionFilter;
 import io.reactivex.rxjava3.core.Flowable;
 
 /**
  * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
  * @author GraviteeSource Team
  */
-public abstract class AbstractFlowResolver implements FlowResolver {
+public abstract class AbstractFlowResolver<T extends BaseExecutionContext> implements FlowResolver<T> {
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-flow/src/main/java/io/gravitee/gateway/reactive/flow/AbstractFlowResolver.java]: just to have a comment :p
```suggestion
public abstract class AbstractFlowResolver<C extends BaseExecutionContext> implements FlowResolver<T> {
```

File: gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/policy/ConditionalPolicy.java [hunk 2]
```diff
@@ -29,16 +29,20 @@
  * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
  * @author GraviteeSource Team
  */
-public class ConditionalPolicy implements Policy, ConditionSupplier {
+public class ConditionalPolicy implements HttpPolicy, ConditionSupplier {
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/policy/ConditionalPolicy.java]: Should we rename this HttpConditionalPolicy to be consistent ?
```suggestion
public class HttpConditionalPolicy implements HttpPolicy, ConditionSupplier {
```

File: gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/policy/DefaultPolicyFactory.java [hunk 3]
```diff
@@ -34,7 +34,7 @@
  */
 public class DefaultPolicyFactory implements PolicyFactory {
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/policy/DefaultPolicyFactory.java]: should it be renamed to `DefaultHttpPolicyFactory` ?

File: gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/v4/policy/DefaultPolicyChainFactory.java [hunk 4]
```diff
@@ -83,7 +83,7 @@ public PolicyChain create(final String flowChainId, Flow flow, ExecutionPhase ph
         if (policyChain == null) {
             final List<Step> steps = getSteps(flow, phase);
 
-            final List<Policy> policies = steps
+            final List<HttpPolicy> policies = steps
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/v4/policy/DefaultPolicyChainFactory.java:86]: should the class be renamed `DefaultHttpPolicyChainFactory` or be generified ?

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/forward/XForwardForProcessor.java [hunk 5]
```diff
@@ -16,10 +16,9 @@
 package io.gravitee.gateway.reactive.reactor.processor.forward;
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/forward/XForwardForProcessor.java:16]: maybe we should put it in a dedicated http package ?

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/metrics/MetricsProcessor.java [hunk 6]
```diff
@@ -53,14 +55,14 @@ public String getId() {
     }
 
     @Override
-    public Completable execute(final MutableExecutionContext ctx) {
+    public Completable execute(final HttpExecutionContextInternal ctx) {
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/metrics/MetricsProcessor.java:58]: (github does not let me comment the `package` line, but my comment is about this class)
should it also be in a dedicated http package ?

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/notfound/NotFoundProcessor.java [hunk 7]
```diff
@@ -44,7 +41,7 @@ public String getId() {
     }
 
     @Override
-    public Completable execute(final MutableExecutionContext ctx) {
+    public Completable execute(final HttpExecutionContextInternal ctx) {
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/notfound/NotFoundProcessor.java:44]: same ?

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/responsetime/ResponseTimeProcessor.java [hunk 8]
```diff
@@ -15,7 +15,7 @@
  */
 package io.gravitee.gateway.reactive.reactor.processor.responsetime;
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/responsetime/ResponseTimeProcessor.java:16]: same question for all the processors

File: gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-shared-policy-group/src/main/java/io/gravitee/gateway/handlers/sharedpolicygroup/policy/SharedPolicyGroupPolicyFactory.java [hunk 9]
```diff
@@ -40,13 +40,16 @@ public boolean accept(PolicyManifest policyManifest) {
...
     @Override
-    protected Policy createPolicy(
+    protected HttpPolicy createPolicy(
         ExecutionPhase phase,
         PolicyManifest policyManifest,
         PolicyConfiguration policyConfiguration,
         PolicyMetadata policyMetadata
     ) {
-        Policy policy = new SharedPolicyGroupPolicy(policyMetadata.getName(), (SharedPolicyGroupPolicyConfiguration) policyConfiguration);
+        HttpPolicy policy = new SharedPolicyGroupPolicy(
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-shared-policy-group/src/main/java/io/gravitee/gateway/handlers/sharedpolicygroup/policy/SharedPolicyGroupPolicyFactory.java:49]: should we have this kind of renaming ?
```suggestion
        HttpPolicy policy = new SharedHttpPolicyGroupPolicy(
```

**Review by ytvnr — APPROVED** [review]

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/forward/XForwardForProcessor.java (hunk 5 above)
Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/forward/XForwardForProcessor.java:16]: I didn't want to add more refactoring at this stage. Can we consider it for later?

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/metrics/MetricsProcessor.java (hunk 6 above)
Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/metrics/MetricsProcessor.java:58]: Same?

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/notfound/NotFoundProcessor.java (hunk 7 above)
Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/notfound/NotFoundProcessor.java:44]: Same? (😅)

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/responsetime/ResponseTimeProcessor.java (hunk 8 above)
Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/responsetime/ResponseTimeProcessor.java:16]: Same counter-question for all the processors 😁

File: gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-shared-policy-group/src/main/java/io/gravitee/gateway/handlers/sharedpolicygroup/policy/SharedPolicyGroupPolicyFactory.java (hunk 9 above)
Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-shared-policy-group/src/main/java/io/gravitee/gateway/handlers/sharedpolicygroup/policy/SharedPolicyGroupPolicyFactory.java:49]: `HttpSharedPolicyGroupPolicy` just to keep the 'Http' wording in the same place?

Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-shared-policy-group/src/main/java/io/gravitee/gateway/handlers/sharedpolicygroup/policy/SharedPolicyGroupPolicyFactory.java:49]: Or maybe we can wait to do that when we will migrate this policy to implement HttpPolicy (it still implements the old Policy interface at the moment)?

File: gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/policy/DefaultPolicyFactory.java (hunk 3 above)
Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/policy/DefaultPolicyFactory.java]: Does the following suit you?

```
public class HttpPolicyFactory implements PolicyFactory<HttpPolicy> {
...
}
```

File: gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/v4/policy/DefaultPolicyChainFactory.java (hunk 4 above)
Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/v4/policy/DefaultPolicyChainFactory.java:86]: I've renamed it HttpPolicyChainFactory to follow the same logic as previously.
I believe there is room for factorization of pieces of code for Kafka version for PolicyChain, PolicyChainFactory, PolicyFactory and PolicyManager

File: gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/policy/DefaultPolicyFactory.java (hunk 3 above)
Comment (ytvnr) [reply gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/policy/DefaultPolicyFactory.java]: yes !

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/forward/XForwardForProcessor.java (hunk 5 above)
Comment (ytvnr) [reply gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactive/reactor/processor/forward/XForwardForProcessor.java:16]: sure, it was just to have this in mind

File: gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-shared-policy-group/src/main/java/io/gravitee/gateway/handlers/sharedpolicygroup/policy/SharedPolicyGroupPolicyFactory.java (hunk 9 above)
Comment (ytvnr) [reply gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-shared-policy-group/src/main/java/io/gravitee/gateway/handlers/sharedpolicygroup/policy/SharedPolicyGroupPolicyFactory.java:49]: makes sense !

**Pull request merged**